Implement simple POST request

christophermanahan opened the pull request:
### Description

- Add the ability to respond to POST requests with an HTTP response that echoes the POST body
- Add a runnable context consumer to the server for easy testing and improved behavior control
- Add Request Parsing and Request Handling
- Utilize Optional for Connection reads to handle exceptions and null values

christophermanahan committed:
Add simple post integration test

christophermanahan committed:
Implement request parsing and handling in HTTPServer
Add top down testing mocks to HTTPServer to drive out the design of request parsing, request building, request handling, and response building

christophermanahan committed:
Add RequestParser

christophermanahan committed:
Add RequestHandler
Add RequestHandler and dependent HTTPResponseBuilder

christophermanahan committed:
Add HTTPResponseBuilder and update HTTPResponse

christophermanahan committed:
Reset state of parser per parse

damonkelley commented on src/main/java/io/github/christophermanahan/carnitas/Constants.java:
```diff
@@ -0,0 +1,7 @@
+package io.github.christophermanahan.carnitas;
+
+public class Constants {
```
Does this need to live on its own? I feel like it might feel more at home as part of another class.

damonkelley commented on src/main/java/io/github/christophermanahan/carnitas/HTTPRequest.java:
```diff
@@ -0,0 +1,19 @@
+        this.body = body;
+    }
+
+    public RequestMethod requestMethod() {
```
Is method named as if we are _requesting the Method_, or is it named because we get a `RequestMethod`? I wonder if just `method` would suffice?

christophermanahan replied:
RequestMethods are removed for now, and string literals are used until an enum may become more useful.

christophermanahan committed:
Add more extensive test coverage to RequestParser
Add tests that drive the parser to reset it's state and check if request is of type POST before attaching body

damonkelley commented on src/main/java/io/github/christophermanahan/carnitas/RequestParser.java:
```diff
@@ -0,0 +1,81 @@
+import java.util.List;
+import java.util.Optional;
+
+public class RequestParser implements Parser {
```
I am wondering what we can do to make this simpler.

christophermanahan replied:
@damonkelley I'm not entirely sure what you mean.

damonkelley replied:
This class still has multiple responsibilities. We should be on the lookout to see how we can simplify it in the future.

damonkelley commented on src/main/java/io/github/christophermanahan/carnitas/RequestParser.java:
```diff
@@ -0,0 +1,81 @@
+
+    public Optional<Request> parse(Receiver receiver) {
+        this.receiver = receiver;
+        resetState();
```
This is typically a smell. Instead of resetting the state, we should look at how we can eliminating mutating the state in the first place.

damonkelley commented on src/main/java/io/github/christophermanahan/carnitas/RequestParser.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    public Optional<Request> parse(Receiver receiver) {
+        this.receiver = receiver;
```
This is also typically a smell. We should strive to only set instance variables in the constructor. There are exceptions to this of course, such as the builder pattern.

damonkelley commented on src/main/java/io/github/christophermanahan/carnitas/StatusCodes.java:
```diff
@@ -0,0 +1,7 @@
+package io.github.christophermanahan.carnitas;
+
+public class StatusCodes {
```
How do you see this evolving? Are status codes coupled to request methods?

christophermanahan replied:
Removed for now and placed into the RequestHandler.

damonkelley commented on src/main/java/io/github/christophermanahan/carnitas/Receiver.java:
```diff
@@ -0,0 +1,6 @@
+package io.github.christophermanahan.carnitas;
+
+public interface Receiver {
```
What is a `Receiver` and why would I want one?

damonkelley replied:
It feels like a very generic name, with a very specific API.

christophermanahan replied:
Removed!

damonkelley commented on src/main/java/io/github/christophermanahan/carnitas/RequestBuilder.java:
```diff
@@ -0,0 +1,7 @@
+package io.github.christophermanahan.carnitas;
+
+public interface RequestBuilder {
```
Do we need a request builder?

christophermanahan replied:
Removed!

christophermanahan committed:
Reword SEND_TO_CONNECTION

damonkelley commented on src/main/java/io/github/christophermanahan/carnitas/RequestMethod.java:
```diff
@@ -0,0 +1,7 @@
+package io.github.christophermanahan.carnitas;
+
+public enum RequestMethod {
```
Does this need to live by itself? I wonder if it would make sense as an inner class?

damonkelley commented on src/main/java/io/github/christophermanahan/carnitas/RequestReceiver.java:
```diff
@@ -0,0 +1,30 @@
+        try {
+            return reader.readLine();
+        } catch (IOException e) {
+            throw new RuntimeException();
```
Is this helpful? Are we going to know what happened if this exception gets thrown?

damonkelley commented on src/main/java/io/github/christophermanahan/carnitas/Receiver.java:
```diff
@@ -0,0 +1,6 @@
+
+public interface Receiver {
+    public String receiveLine();
+    public String receiveCharacters(int amount);
```
I am wondering why we need both of these methods.

damonkelley commented on src/main/java/io/github/christophermanahan/carnitas/ResponseBuilder.java:
```diff
@@ -0,0 +1,8 @@
+package io.github.christophermanahan.carnitas;
+
+public interface ResponseBuilder {
```
Do we need a response builder?

christophermanahan replied:
Removed!

damonkelley commented on src/main/java/io/github/christophermanahan/carnitas/Handler.java:
```diff
@@ -0,0 +1,5 @@
+package io.github.christophermanahan.carnitas;
+
+public interface Handler {
```
Nice!

damonkelley commented on src/main/java/io/github/christophermanahan/carnitas/HTTPServer.java:
```diff
@@ -21,8 +24,8 @@ void run() {
-        connection.receive()
-          .map(request -> new HTTPResponse())
+        parser.parse(connection.receiver())
+          .map(handler::handle)
```
Nice!

damonkelley commented on src/main/java/io/github/christophermanahan/carnitas/RequestParser.java:
```diff
@@ -0,0 +1,81 @@
+          .body(body)
+          .build();
+    }
+}
```
It feels like this class has multiple responsibilities

damonkelley commented on src/test/java/io/github/christophermanahan/carnitas/HTTPRequestTest.java:
```diff
@@ -0,0 +1,20 @@
+    @Test
+    void getsRequestMethod() {
+        RequestMethod requestMethod = RequestMethod.GET;
+        Assertions.assertEquals(requestMethod, new HTTPRequest(requestMethod, null).requestMethod());
```
I tend to avoid using `null`, even if it is in a test. I prefer to use something that resembles production data. `null` does it give me any context.

damonkelley commented on src/test/java/io/github/christophermanahan/carnitas/HTTPResponseBuilderTest.java:
```diff
@@ -0,0 +1,41 @@
+          + Headers.CONTENT_LENGTH + body.length()
+          + Constants.BLANK_LINE
+          + body;
+        assertEquals(expectedResponse, new String(response.serialize()));
```
I wonder if there is an easier way to test this instead of going through `serialize`

christophermanahan replied:
@damonkelley I haven't figured that out yet, it pops up in other tests as well.

damonkelley commented on src/test/java/io/github/christophermanahan/carnitas/HTTPResponseBuilderTest.java:
```diff
@@ -0,0 +1,41 @@
+          + body;
+        assertEquals(expectedResponse, new String(response.serialize()));
+    }
+}
```
Missing a new line at the end of the file.

damonkelley commented on src/test/java/io/github/christophermanahan/carnitas/HTTPServerTest.java:
```diff
@@ -39,56 +43,70 @@ void connectionIsClosedAfterResponse() {
+                connections.next();
+                throw new RuntimeException(ErrorMessages.ACCEPT_CONNECTION);
+            }
+        }
```
This isn't exactly what I had in mind, but it works. 😄

damonkelley commented on src/test/java/io/github/christophermanahan/carnitas/RequestParserTest.java:
```diff
@@ -0,0 +1,156 @@
+        Optional<Request> parsed = parser.parse(receiver);
+
+        String expectedRequest = requestMethod + body;
+        String parsedRequest = parsed.isEmpty() ? null : parsed.get().body();
```
I tend to think of if statements in tests as being a smell.

damonkelley commented on src/test/java/io/github/christophermanahan/carnitas/RequestParserTest.java:
```diff
@@ -0,0 +1,156 @@
+        }
+    }
+
+    private class TestRequest implements Request {
```
`HTTPRequest` doesn't have any behavior, does it? If not, you could probably use the real thing instead of a `TestRequest`

damonkelley commented on src/test/java/io/github/christophermanahan/carnitas/RequestReceiverTest.java:
```diff
@@ -0,0 +1,73 @@
+    @Test
+    void receivesCharacterAmountFromStream() {
+        String request = "GET /simple_get HTTP1.1";
+        int amount = 3;
```
amount of what?

damonkelley commented on src/test/java/io/github/christophermanahan/carnitas/RequestReceiverTest.java:
```diff
@@ -0,0 +1,73 @@
+
+        Receiver receiver = new RequestReceiver(new ReadLineExceptionReader(new StringReader(request)));
+
+        assertThrows(RuntimeException.class, receiver::receiveLine);
```
It might be different in JUnit 5, but I think there may also be a way to do this using an annotation. FWIW

damonkelley commented on src/test/java/io/github/christophermanahan/carnitas/RequestReceiverTest.java:
```diff
@@ -0,0 +1,73 @@
+
+        Receiver receiver = new RequestReceiver(new ReadExceptionReader(new StringReader(request)));
+
+        assertThrows(RuntimeException.class, ()->{ receiver.receiveCharacters(request.length()); } );
```
I think the curly braces are redundant here.

damonkelley commented on src/test/resources/gradle/cucumber/multipleRequests.feature:
```diff
@@ -3,6 +3,6 @@ Feature: Multiple Requests
   Scenario: Multiple requests
     Given The server is running on port "33333"
-    When I send method "GET" for "/simple_get" to host at the specified port 3 times
+    When I send method "GET" for "/simple_get" to host at the specified port 1 time
```
Do we need this test?

damonkelley replied:
Maybe we could just break this up into multple steps instead of specifying how many times the same request is sent. Repeating the same request doesn't seem like a typical real world scenario.

christophermanahan replied:
@damonkelley I will work on modifying these acceptance tests in the next iteration.

christophermanahan replied:
Had some extra time to modify them and implement this suggestion!

damonkelley commented on src/test/resources/gradle/cucumber/simplePost.feature:
```diff
@@ -0,0 +1,9 @@
+
+  Scenario: Simple POST request
+    Given The server is running on port "33333"
+    When I send method "POST" for "/simple_post" with body "hello world" to host at the specified port 1 time
```
The `1 time` thing feels like overkill tot me.

christophermanahan replied:
See my response to your other comment on feature tests.

damonkelley commented on src/test/java/io/github/christophermanahan/carnitas/HTTPServerTest.java:
```diff
@@ -130,36 +138,90 @@ public TestConnection accept() {
-    private class CloseException extends TestConnection {
-        CloseException(String request) {
-            super(request);
+    private class TestRequest implements Request {
```
See my other comment about using `HTTPRequest`

damonkelley commented on src/test/java/io/github/christophermanahan/carnitas/HTTPServerTest.java:
```diff
@@ -130,36 +138,90 @@ public TestConnection accept() {
+        }
+    }
+
+    private class TestResponse implements Response {
```
Do we need this? Could we use `HTTPResponse` instead?

damonkelley commented on src/test/java/io/github/christophermanahan/carnitas/HTTPServerTest.java:
```diff
@@ -130,36 +138,90 @@ public TestConnection accept() {
-    private class SendException extends TestConnection {
-        SendException(String request) {
-            super(request);
+    private class TestReceiver implements Receiver {
```
This feels like overkill. I think it is a symptom of the `Connection` depending on the a `Receiver` and exposing it in the API. What can we do about that?

christophermanahan replied:
Change the API to be consistent with a connection abstraction!

damonkelley reviewed (commented):
Lots of good stuff in here! Also some complexity. Let's see what we can do to eliminate that!

christophermanahan committed:
Rename Acceptor to Listener and add ConnectionReader

christophermanahan committed:
Add WhileOpen RunContext

christophermanahan committed:
Rewrite HTTPServer in parallel
Rewrite HTTPServer to use a run context to control serving connections and a ConnectionReader to obviate the receiver and simplify testing

avni510 commented on src/main/java/io/github/christophermanahan/carnitas/HTTPResponseBuilder.java:
```diff
@@ -0,0 +1,28 @@
+
+public class HTTPResponseBuilder implements ResponseBuilder {
+    private String statusCode = "";
+    private String version = "";
```
You may want to move some of those constants here if it feels right

christophermanahan replied:
I ended up moving the constants into the HTTPResponse after removing the builders!

avni510 commented on src/main/java/io/github/christophermanahan/carnitas/Main.java:
```diff
@@ -4,23 +4,25 @@
+    private static Listener connectionListener;
 
     public static void main(String[] args) {
         int port = args.length == 0 ? 33333 : Integer.parseInt(args[0]);
```
It feels like this should be wrapped in a function to make it more descriptive as to what's going on

christophermanahan replied:
I have a story that will obviate the need for this field that I will be working on today!

avni510 commented on src/main/java/io/github/christophermanahan/carnitas/Request.java:
```diff
@@ -0,0 +1,6 @@
+package io.github.christophermanahan.carnitas;
+
+public interface Request {
```
I'm not completely sure why `Request` and `RequestBuilder` have interfaces

christophermanahan replied:
The interfaces have been removed.

avni510 commented on src/test/java/gradle/cucumber/Client.java:
```diff
@@ -8,16 +8,24 @@
         HttpRequest request = HttpRequest.newBuilder()
             .uri(URI.create("http://localhost:" + port + location))
-            .method(method, HttpRequest.BodyPublishers.noBody())
+            .version(HttpClient.Version.HTTP_1_1)
```
I think setting the version makes sense to keep in the builder as it rarely changes for specific requests

avni510 commented on src/test/java/io/github/christophermanahan/carnitas/HTTPResponseBuilderTest.java:
```diff
@@ -0,0 +1,41 @@
+
+        Response response = new HTTPResponseBuilder()
+          .statusCode(statusCode)
+          .version(version)
```
I think it makes sense to set the version inside the builder as the version rarely changes

christophermanahan replied:
The version is now located in the HTTPResponse class.

christophermanahan committed:
Refactor Connection to extend Readable

christophermanahan committed:
Add stateless RequestParser

christophermanahan committed:
Refactor RequestParser2 to handle empty reads
Change Reader to return Optionals to prevent throwing exceptions on failed reads. Change HTTPRequest2 to have an Optional body.

christophermanahan committed:
Add RequestHandler2 and modify HTTPResponse2
HTTPResponse2 now includes a withBody method that accepts an optional body with empty default

christophermanahan committed:
Alter Connection and ConnectionReader to return Optionals
Failed socket reads now return empty optionals instead of throwing exceptions

christophermanahan committed:
Edit RequestParser2
Edit RequestParser 2 so that contentLength is not overwritten when multiple headers are present

christophermanahan committed:
Parallel refactor main to use the new HTTPServer

christophermanahan committed:
Remove unused methods from Listener interface

christophermanahan committed:
Remove unused StatusCodes class

christophermanahan committed:
Move Constants into HTTPResponse

christophermanahan committed:
Move status codes into RequestHandler

christophermanahan committed:
Remove number of requests from acceptance tests

damonkelley commented on src/test/java/io/github/christophermanahan/carnitas/HTTPServerTest.java:
```diff
@@ -3,163 +3,180 @@
+            public Optional<Character> read() {
+                index++;
+                return Optional.of(
+                  List.of(HTTPResponse.CRLF.split("")).get(index).charAt(0)
```
Is this necessary?

christophermanahan replied:
No, mocking the parser negates the need for this.

damonkelley commented on src/test/java/io/github/christophermanahan/carnitas/HTTPServerTest.java:
```diff
@@ -3,163 +3,180 @@
-        public TestConnection accept() {
-            return connections.next();
+        public Optional<Character> read() {
+            index++;
```
Is there any way for us to avoid using indexes like this?

christophermanahan replied:
Yes, switched over to an iterator!

damonkelley commented on src/test/java/io/github/christophermanahan/carnitas/RequestParserTest.java:
```diff
@@ -0,0 +1,87 @@
+        }
+
+        public Optional<String> readUntil(String delimiter) {
+            index++;
```
Same question here

christophermanahan replied:
Same answer here!

christophermanahan committed:
Add .iml files to .gitignore

christophermanahan committed:
Refactor HTTPServerTest Connection to use an iterator

christophermanahan committed:
Refactor RequestParserTest Reader to use an iterator

christophermanahan committed:
Refactor HTTPServerTest to mock the parser for send failure

christophermanahan committed:
Refactor ErrorLoggerTest to strip the output
